common/spec/dependabot/file_fetchers/base_spec.rb

```diff
@@ -1313,4 +1319,79 @@ def fetch_files
       end
     end
   end
+
+  context "with repo_contents_path" do
+    let(:repo_contents_path) { Dir.mktmpdir }
+    after { FileUtils.rm_rf(repo_contents_path) }
+
+    before do
+      allow(file_fetcher_instance).to receive(:commit).and_return("sha")
+    end
+
+    # lazily spoof `git clone` on _clone_repo_contents
+    let(:contents) { fixture("ruby", "gemfiles", "Gemfile") }
+    let(:fill_repo) {}
+    let(:mock_clone) do
+      `git init #{repo_contents_path}`
+      fill_repo
+      `git -C #{repo_contents_path} add .`
+      `git -C #{repo_contents_path} commit --allow-empty -m'fake clone'`
+      repo_contents_path
+    end
+    before do
+      allow(file_fetcher_instance).
+        to receive(:_clone_repo_contents).and_return(mock_clone)
```

Comment:
Slightly worried about stubbing the system under test, we could alternatively put this [in the `tmp` path expected by the FileFetcher](https://github.com/dependabot/dependabot-core/blob/main/common/lib/dependabot/file_fetchers/base.rb#L433), but it does make some assumptions about the implementation, so I'm not sure if it's a net win. Might also be an opportunity to extract an object (`RepoCloner` ?) that we could inject.

No need to change anything just wanted to bring it up and make sure we're OK stubbing the class being tested.

Reply:
Alternatively, we could inject a `file://` based `source.url` and actually invoke the `git clone`.

That's how I did tests in $unlinkable - a local "upstream" + "downstream" clone for end-to-end. WDYT about that?

Reply:
I like it! @jasonrudolph and I discussed something similar for #2551 yesterday